File: solutions/1.py
# https://leetcode.com/problems/two-sum/

from typing import List

class Solution:
    # O(n) runtime, O(n) space
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # store previously seen nums and their index
        seen = {}
        for i, x in enumerate(nums):
            # check if the difference needed to add up to the target has been seen yet
            diff = target - x
            if diff in seen:
                return [i, seen[diff]]
            
            seen[x] = i
        return [-1, -1]
        
    # O(n^2) runtime, O(1) space
    def bruteForceSolution(self, nums: List[int], target: int) -> List[int]:
        # iterate through all possible pairs to find solution
        for i in range(len(nums)):
            for j in range(i, len(nums)):
                if nums[i] + nums[j] == target:
                    return [i, j]
        return [-1, -1]

File: solutions/1.ts
// https://leetcode.com/problems/two-sum/

// O(n) runtime, O(n) space
function twoSum(nums: number[], target: number): number[] {
    // keep track of values and their indexes we've seen
    const seen = [];

    for (let i = 0; i < nums.length; i++) {
        const diff = target - nums[i];

        // check if we've seen the difference between the num and target
        if (seen[diff] != undefined) {
            return [i, seen[diff]];
        }
        
        seen[nums[i]] = i;
    }
    
    return [0,0]
};

// Iterate through all possible combinations - O(n^2)
function bruteForceSolution(nums: number[], target: number): number[] {
    for (let i = 0; i < nums.length - 1; i++) {
        for (let j = i+1; j < nums.length; j++) {
            if (nums[i] + nums[j] === target) {
                return [i, j];
            }
        }
    }
}; 